drivers/sensor/grow_r502a/grow_r502a.c

```diff
@@ -0,0 +1,753 @@
+{
+	struct grow_r502a_data *drv_data = dev->data;
+	uint8_t data[5] = { FPS_DELETE, id >> 8, id & 0xFF, count >> 8, count & 0xFF };
+
+	send_packet(dev, FPS_COMMAND_PACKET, data, sizeof(data));
+
+	if (drv_data->buffer[FPS_PID_IDX] != FPS_ACK_PACKET) {
+		LOG_ERR("Error receiving ack packet 0x%X", drv_data->buffer[FPS_PID_IDX]);
+		return -EIO;
+	}
+
+	if (drv_data->buffer[FPS_CC_IDX] == FPS_OK) {
+		LOG_INF("Fingerprint Deleted from ID #%d to #%d", id, (id + count));
+	} else {
+		LOG_ERR("Error deleting image 0x%X", drv_data->buffer[FPS_CC_IDX]);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int fps_empty_db(const struct device *dev)
+{
+	struct grow_r502a_data *drv_data = dev->data;
+	uint8_t data = FPS_EMPTYLIBRARY;
+
+	send_packet(dev, FPS_COMMAND_PACKET, &data, sizeof(data));
+
+	if (drv_data->buffer[FPS_PID_IDX] != FPS_ACK_PACKET) {
+		LOG_ERR("Error receiving ack packet 0x%X", drv_data->buffer[FPS_PID_IDX]);
+		return -EIO;
+	}
+
+	if (drv_data->buffer[FPS_CC_IDX] == FPS_OK) {
+		LOG_INF("Emptied Fingerprint Library");
+	} else {
+		LOG_ERR("Error emptying fingerprint library 0x%X", drv_data->buffer[FPS_CC_IDX]);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int fps_search(const struct device *dev, uint8_t slot,
+				uint16_t *finger_id, uint16_t *matching_score)
+{
+	struct grow_r502a_data *drv_data = dev->data;
+	struct led_params led_ctrl = {
+		.color_code = LED_CTRL_BREATHING,
+		.color = LED_COLOR_BLUE,
+		.speed = LED_SPEED_HALF,
+		.cycle = 0x01,
+	};
+	uint8_t data[6] = { FPS_SEARCH,
+			    slot,
+			    0x00,
+			    0x00,
+			    (uint8_t)(FPS_DEFAULT_CAPACITY >> 8),
+			    (uint8_t)(FPS_DEFAULT_CAPACITY & 0xFF) };
+	int ret;
+
+	send_packet(dev, FPS_COMMAND_PACKET, data, sizeof(data));
+
+	if (drv_data->buffer[FPS_PID_IDX] != FPS_ACK_PACKET) {
+		LOG_ERR("Error receiving ack packet 0x%X", drv_data->buffer[FPS_PID_IDX]);
+		return -EIO;
+	}
+
+	*finger_id = sys_get_be16(&drv_data->buffer[FPS_HEADER_LEN + 1]);
+
+	*matching_score = sys_get_be16(&drv_data->buffer[FPS_HEADER_LEN + 3]);
+
+	if (drv_data->buffer[FPS_CC_IDX] == FPS_OK) {
+		led_ctrl.color_code = LED_CTRL_FLASHING;
+		led_ctrl.color = LED_COLOR_PURPLE;
+		led_ctrl.cycle = 0x01;
+		ret = fps_led_control(dev, &led_ctrl);
+		if (ret != 0) {
+			return ret;
+		}
+		LOG_INF("Found a matching print! at ID #%d", *finger_id);
+	} else if (ret == FPS_NOT_FOUND) {
+		led_ctrl.color_code = LED_CTRL_BREATHING;
+		led_ctrl.color = LED_COLOR_RED;
+		led_ctrl.cycle = 0x02;
+		ret = fps_led_control(dev, &led_ctrl);
+		if (ret != 0) {
+			return ret;
+		}
+		LOG_ERR("Did not find a match");
+	} else {
+		led_ctrl.color_code = LED_CTRL_ON_ALWAYS;
+		led_ctrl.color = LED_COLOR_RED;
+		ret = fps_led_control(dev, &led_ctrl);
+		if (ret != 0) {
+			return ret;
+		}
+		LOG_ERR("Error searching for image 0x%X", drv_data->buffer[FPS_CC_IDX]);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int fps_enroll(const struct device *dev, const struct sensor_value *val)
```

Comment:
All these sequence should be atomic, because the data manipulation of each command are stored in temporary buffer inside chip. Please use locks

Reply:
fixed